response.go

```diff
@@ -294,67 +534,167 @@ func (resp *Response) decodeBodyTyped(res interface{}) (err error) {
+		}
 	}
-	return fmt.Sprintf("<%d ERR 0x%x %s>", resp.RequestId, resp.Code, resp.Error)
+	return err
 }
 
-// Tuples converts result of Eval and Call to same format
-// as other actions returns (i.e. array of arrays).
-func (resp *Response) Tuples() (res [][]interface{}) {
-	res = make([][]interface{}, len(resp.Data))
-	for i, t := range resp.Data {
-		switch t := t.(type) {
-		case []interface{}:
-			res[i] = t
-		default:
-			res[i] = []interface{}{t}
+func (resp *ExecuteResponse) DecodeTyped(res interface{}) error {
+	resp.decodedTyped = true
+
+	var err error
+	if resp.buf.Len() > 0 {
+		offset := resp.buf.Offset()
+		defer resp.buf.Seek(offset)
+
+		info := &decodeInfo{}
+		var l int
+
+		d := msgpack.NewDecoder(&resp.buf)
+		d.SetMapDecoder(func(dec *msgpack.Decoder) (interface{}, error) {
+			return dec.DecodeUntypedMap()
+		})
+
+		if l, err = d.DecodeMapLen(); err != nil {
+			return err
 		}
+		for ; l > 0; l-- {
+			var cd int
+			if cd, err = smallInt(d, &resp.buf); err != nil {
+				return err
+			}
+			decoded, err := decodeTypedCommonField(d, res, cd, info)
+			if err != nil {
+				return err
+			}
+			if !decoded {
+				switch iproto.Key(cd) {
+				case iproto.IPROTO_SQL_INFO:
+					if err = d.Decode(&resp.sqlInfo); err != nil {
+						return err
+					}
+				case iproto.IPROTO_METADATA:
+					if err = d.Decode(&resp.metaData); err != nil {
+						return err
+					}
+				default:
+					if err = d.Skip(); err != nil {
+						return err
+					}
+				}
+			}
+		}
+		if info.decodedError != "" {
+			resp.header.Error = iproto.Error(int(resp.header.Error) &^
+				int(iproto.IPROTO_TYPE_ERROR))
+			err = Error{resp.header.Error, info.decodedError, info.errorExtendedInfo}
+		}
+	}
+	return err
+}
+
+func (resp *baseResponse) Header() Header {
+	return resp.header
+}
+
+// Pos returns a position descriptor of the last selected tuple for the SelectResponse.
+// If the response was not decoded, this method will call Decode().
+func (resp *SelectResponse) Pos() ([]byte, error) {
+	if !resp.decoded && !resp.decodedTyped {
+		resp.Decode()
+	}
+	return resp.pos, resp.err
+}
+
+// MetaData returns ExecuteResponse meta-data.
+// If the response was not decoded, this method will call Decode().
+func (resp *ExecuteResponse) MetaData() ([]ColumnMetaData, error) {
+	if !resp.decoded && !resp.decodedTyped {
+		resp.Decode()
+	}
+	return resp.metaData, resp.err
+}
+
+// SQLInfo returns ExecuteResponse sql info.
+// If the response was not decoded, this method will call Decode().
+func (resp *ExecuteResponse) SQLInfo() (SQLInfo, error) {
+	if !resp.decoded && !resp.decodedTyped {
+		resp.Decode()
+	}
+	return resp.sqlInfo, resp.err
+}
+
+// String implements Stringer interface.
+func (resp *baseResponse) String() (str string) {
+	if resp.header.Error == ErrorNo {
+		return fmt.Sprintf("<%d OK %v>", resp.header.RequestId, resp.data)
 	}
-	return res
+	return fmt.Sprintf("<%d ERR %x %v>", resp.header.RequestId, resp.header.Error, resp.err)
```

Comment:
iproto.Error has a `String()` method.
```suggestion
	return fmt.Sprintf("<%d ERR %s %v>", resp.header.RequestId, resp.header.Error, resp.err)
```